spark/src/main/java/org/apache/iceberg/spark/actions/BaseRepairManifestsSparkAction.java

```diff
@@ -0,0 +1,313 @@
+    StructType sparkType = (StructType) manifestEntryDf.schema().apply("data_file").dataType();
+    JavaRDD<Row> manifestEntryRdd = manifestEntryDf.toJavaRDD();
+    JavaPairRDD<ManifestFileInfo, Iterable<Row>> entriesByManifest = manifestEntryRdd.groupBy(
+        r -> new ManifestFileInfo(r.getString(0), r.getInt(1)));
+
+    // Calculate manifest entries for repair
+    RepairManifestHelper.RepairOptions options = new RepairManifestHelper.RepairOptions(repairMetrics);
+    Broadcast<Table> broadcastTable = sparkContext().broadcast(SerializableTable.copyOf(getTable()));
+    Broadcast<SerializableConfiguration> conf = sparkContext().broadcast(hadoopConf);
+    JavaRDD<Tuple2<ManifestFileInfo, Iterable<PatchedManifestEntry>>> toRepair =
+        entriesByManifest.map(calculateRepairs(broadcastTable, conf, sparkType, options))
+            // Process only manifest files with repaired entries
+        .filter(m -> StreamSupport.stream(m._2.spliterator(), false)
+            .anyMatch(p -> p.repaired));
+
+    // Write out repaired manifests
+    Broadcast<FileIO> io = sparkContext().broadcast(getFileIO());
+    JavaRDD<ManifestFile> repairedManifests = toRepair.map(
+        writeRepairs(io, broadcastTable, formatVersion, stagingLocation));
+
+    // Prepare results
+    List<ManifestFile> addedManifests = repairedManifests.collect();
+    List<ManifestFile> deletedManifests = toRepair.collect().stream().map(t -> {
+      String path = t._1().getPath();
+      ManifestFile mf = manifestFiles.get(path);
+      // Sanity check deleted file existed in original list
+      Preconditions.checkNotNull(mf, "Manifest file cannot be null for " + path);
+      return mf;
+    }).collect(Collectors.toList());
+
+    Iterable<ManifestFile> newManifests = replaceManifests(deletedManifests, addedManifests);
+    return new BaseRepairManifestsActionResult(deletedManifests, Lists.newArrayList(newManifests));
+  }
+
+
+  private Map<String, ManifestFile> findMatchingManifests() {
+    Snapshot currentSnapshot = getTable().currentSnapshot();
+    if (currentSnapshot == null) {
+      return ImmutableMap.of();
+    }
+    return currentSnapshot.dataManifests().stream()
+        .filter(manifest -> predicate.test(manifest))
+        .collect(Collectors.toMap(ManifestFile::path, mf -> mf));
+  }
+
+  private Dataset<Row> buildManifestEntryDF(List<ManifestFile> manifests) {
+    Dataset<Row> manifestDF = spark()
+        .createDataset(Lists.transform(manifests, m -> new ManifestFileInfo(m.path(), m.partitionSpecId())),
+            Encoders.bean(ManifestFileInfo.class))
+        .toDF("partition_spec_id", "manifest_path");
+    Dataset<Row> manifestEntryDF = loadMetadataTable(getTable(), ENTRIES)
+        .filter("status < 2") // select only live entries
+        .selectExpr("input_file_name() as manifest_path", "snapshot_id", "sequence_number", "data_file");
+
+    return manifestEntryDF.as("manifest_entry")
```

Comment:
I still think we should fix this in the actual ManfiestEntry Row and have the spec generated at read time, it feels a little odd to me to be joining to reconnect it when we knew the right ID when we read it the first time. I'm fine with this for now but we really need to get SpecID into that metadata table.

Reply:
I see, I will file a ticket for that if none exist.  What is the policy for adding to metadata table schemas, ie do you know if we have some backward-compatibility policy?